File: src/main/java/org/processmining/models/time_driven_behavior/ResourceMapping.java
package org.processmining.models.time_driven_behavior;

import org.processmining.models.organizational_extension.Group;
import org.processmining.models.organizational_extension.Resource;
import org.processmining.models.organizational_extension.Role;

import java.util.*;

/**
 * @author Ivan Shugurov
 *         Created on 17.04.2014
 */
public class ResourceMapping
{
    private List<Group> selectedGroups = new ArrayList<Group>();
    private List<Role> selectedRoles = new ArrayList<Role>();
    private List<Resource> selectedResources = new ArrayList<Resource>();
    private List<Resource> selectedSimplifiedResources = new ArrayList<Resource>();

    public void addSelectedGroup(Group selectedGroup)
    {
        selectedGroups.add(selectedGroup);
    }

    public void removeSelectedGroup(Group selectedGroup)
    {
        selectedGroups.remove(selectedGroup);
        for (Role role : selectedGroup.getRoles())
        {
            removeSelectedRole(role);
        }
    }

    public void addSelectedRole(Role selectedRole)
    {
        selectedRoles.add(selectedRole);
    }

    public void removeSelectedRole(Role selectedRole)
    {
        boolean wasRemoved = selectedRoles.remove(selectedRole);
        if (wasRemoved)
        {
            for (Resource resource : selectedRole.getResources())
            {
                removeSelectedResources(resource);
            }
        }
    }

    private void removeAllSelectedRoles(Collection<Role> roles)
    {
        selectedRoles.removeAll(roles);
        for (Role role : roles)
        {
            removeAllSelectedResources(role.getResources());
        }
    }

    public void addSelectedResource(Resource selectedResource)
    {
        selectedResources.add(selectedResource);
    }

    public void removeSelectedResources(Resource selectedResource)
    {
        selectedResources.remove(selectedResource);
    }

    public void addSelectedSimplifiedResource(Resource selectedSimplifiedResource)
    {
        selectedSimplifiedResources.add(selectedSimplifiedResource);
    }

    public void removeSelectedSimplifiedResource(Resource simplifiedResource)
    {
        selectedSimplifiedResources.remove(simplifiedResource);
    }


    public List<Group> getSelectedGroups()
    {
        return Collections.unmodifiableList(selectedGroups);
    }

    public List<Role> getSelectedRoles()
    {
        return Collections.unmodifiableList(selectedRoles);
    }

    public List<Resource> getSelectedResources()
    {
        return Collections.unmodifiableList(selectedResources);
    }

    public List<Resource> getSelectedSimplifiedResources()
    {
        return Collections.unmodifiableList(selectedSimplifiedResources);
    }

    public void retainSelectedGroups(Collection<Group> groups)
    {
        Iterator<Group> groupIterator = selectedGroups.iterator();
        while (groupIterator.hasNext())
        {
            Group group = groupIterator.next();
            if (!groups.contains(group))
            {
                groupIterator.remove();
                removeAllSelectedRoles(group.getRoles());
            }
        }
    }

    private void removeAllSelectedResources(Collection<Resource> resources)
    {
        selectedResources.removeAll(resources);
        selectedSimplifiedResources.removeAll(resources);
    }

    public void retainSelectedRoles(Collection<Role> roles)
    {
        Iterator<Role> roleIterator = selectedRoles.iterator();
        while (roleIterator.hasNext())
        {
            Role role = roleIterator.next();
            if (!roles.contains(role))
            {
                roleIterator.remove();
                removeAllSelectedResources(role.getResources());
            }
        }
    }

    public void retainSelectedResources(Collection<Resource> resources)
    {
        selectedResources.retainAll(resources);
        selectedSimplifiedResources.retainAll(resources);
    }

}

File: console/src/main/kotlin/org/processmining/models/time_driven_behavior/ResourceMapping.kt
package org.processmining.models.time_driven_behavior

import org.processmining.models.organizational_extension.Group
import org.processmining.models.organizational_extension.Resource
import org.processmining.models.organizational_extension.Role

import java.util.*

/**
 * @author Ivan Shugurov
 * Created on 17.04.2014
 */
class ResourceMapping(
//        val selectedGroups: List<Group>,
//        val selectedRoles : List<Role>,
        val selectedResources : List<Resource> = emptyList(),
        val selectedSimplifiedResources : List<Resource> = emptyList()
) {
    init {
//        if (selectedResources.size + selectedSimplifiedResources.size < 1)
//            throw IllegalArgumentException("Transition should have at list one resource in ResourceMapping.")
    }
    
//    private val _selectedGroups = mutableListOf<Group>()
//    private val _selectedRoles = mutableListOf<Role>()
//    private val _selectedResources = mutableListOf<Resource>()
//    private val _selectedSimplifiedResources = mutableListOf<Resource>()
    
//    fun addSelectedGroup(selectedGroup: Group) {
//        _selectedGroups.add(selectedGroup)
//    }
//    
//    fun removeSelectedGroup(selectedGroup: Group) {
//        _selectedGroups.remove(selectedGroup)
//        for (role in selectedGroup.roles) {
//            removeSelectedRole(role)
//        }
//    }
//    
//    fun addSelectedRole(selectedRole: Role) {
//        _selectedRoles.add(selectedRole)
//    }
//    
//    fun removeSelectedRole(selectedRole: Role) {
//        val wasRemoved = _selectedRoles.remove(selectedRole)
//        if (wasRemoved) {
//            for (resource in selectedRole.resources) {
//                removeSelectedResources(resource)
//            }
//        }
//    }
//    
//    private fun removeAllSelectedRoles(roles: Collection<Role>) {
//        _selectedRoles.removeAll(roles)
//        for (role in roles) {
//            removeAllSelectedResources(role.resources)
//        }
//    }
//    
//    fun addSelectedResource(selectedResource: Resource) {
//        _selectedResources.add(selectedResource)
//    }
//    
//    fun removeSelectedResources(selectedResource: Resource) {
//        _selectedResources.remove(selectedResource)
//    }
//    
//    fun addSelectedSimplifiedResource(selectedSimplifiedResource: Resource) {
//        _selectedSimplifiedResources.add(selectedSimplifiedResource)
//    }
//    
//    fun removeSelectedSimplifiedResource(simplifiedResource: Resource) {
//        _selectedSimplifiedResources.remove(simplifiedResource)
//    }
//    
//    
//    val selectedGroups: List<Group>
//        get() = _selectedGroups
//    
//    val selectedRoles: List<Role>
//        get() = _selectedRoles
//    
//    val selectedResources: List<Resource>
//        get() = _selectedResources
//    
//    val selectedSimplifiedResources: List<Resource>
//        get() = _selectedSimplifiedResources
//    
//    fun retainSelectedGroups(groups: Collection<Group>) {
//        val groupIterator = _selectedGroups.iterator()
//        while (groupIterator.hasNext()) {
//            val group = groupIterator.next()
//            if (!groups.contains(group)) {
//                groupIterator.remove()
//                removeAllSelectedRoles(group.roles)
//            }
//        }
//    }
//    
//    private fun removeAllSelectedResources(resources: Collection<Resource>) {
//        _selectedResources.removeAll(resources)
//        _selectedSimplifiedResources.removeAll(resources)
//    }
//    
//    fun retainSelectedRoles(roles: Collection<Role>) {
//        _selectedRoles
//                .filter { it !in roles }
//                .forEach { removeAllSelectedResources(it.resources) }
//        
////        val roleIterator = _selectedRoles.iterator()
////        while (roleIterator.hasNext()) {
////            val role = roleIterator.next()
////            if (!roles.contains(role)) {
////                roleIterator.remove()
////                removeAllSelectedResources(role.resources)
////            }
////        }
//    }
//    
//    fun retainSelectedResources(resources: Collection<Resource>) {
//        _selectedResources.retainAll(resources)
//        _selectedSimplifiedResources.retainAll(resources)
//    }
    
}
